### PR TITLE
soc/interconnect/wishbone: Add incrementing address burst mode to SRAM

### PR DESCRIPTION
**Pull request by koluckirafal**
This pull request adds support for incrementing address burst cycles to SRAM module. Both linear and wrapped increments are supported.
Because for some designs burst might be unnecessary (e.g. no cache in CPU core), it is disabled by default. It can be turned on by passing parameter `burst=True` in `SoCCore.add_ram()` function, with shortcut available for integrated *RAM/ROM modules in the form of `integrated_{rom,sram,main_ram}_burst` parameters in `SoCCore` class.

Implementation was tested by writing a test suite and an example SoC for measuring speed difference: https://github.com/antmicro/wishbone-interconnect-burst-mode-benchmark (README contains results and instructions how to run tests).

Wishbone burst cycle documentation: https://wishbone-interconnect.readthedocs.io/en/latest/04_registered.html

### CODE REVIEW

**Review by enjoy-digital — CHANGES_REQUESTED** [review]
Thanks @koluckirafal. I just did a quick first review:
- Burst support should probably be enabled at the bus level (through `bus_bursting `parameter that would then be propagated to memories).
- To be able to maintain this feature we'll need a unit-tests that cover the different burst mode/corner cases and would be run in LiteX-CI. Would it be possible to add it?

**Comment by koluckirafal** [comment]
> * Burst support should probably be enabled at the bus level (through `bus_bursting `parameter that would then be propagated to memories).

Done, burst cycles support can be now enabled with `bus_bursting` optional parameter in SoCBusHandler. Corresponding argument in `soc_core_args()` is also now available.

> * To be able to maintain this feature we'll need a unit-tests that cover the different burst mode/corner cases and would be run in LiteX-CI. Would it be possible to add it?

I couldn't find any tests in LiteX for Wishbone modules other than `{Up,Down}Converter`. Should I write tests only for SRAM or also for all other Wishbone modules?

**Comment by enjoy-digital** [comment]
Thanks for the update. The unit-tests are indeed minimal on Wishbone and that's clearly something we should improve in the future. I'm of course not asking you to write unit-tests for the other Wishbone modules but would like to have a unit-test for the SRAM that exercise the burst support (to be able to rework/maintain this code more easily in the future) Thanks!

**Comment by koluckirafal** [comment]
I added a basic unit test for SRAM write and read operations, should be enough for testing incrementing address burst cycle support - currently only linear and 4-beat wrap increments are tested.

**Comment by enjoy-digital** [comment]
Thanks for the update. I added a few review comments in the code. We should  really try to simplify it when possible (otherwise it's difficult to maintenance/support). Can you also add a test for Constant burst? After addressing this it will be fine and I'll do the other improvement I can see myself. Thanks!

**Comment by koluckirafal** [comment]
Test for constant address burst cycle added - it works despite the lack of constant address burst support in wishbone.SRAM, because unsupported cycle types are treated as classic cycle.
About review comments, I couldn't find them, probably due to rebasing I did previously.

**Comment by enjoy-digital** [comment]
Thanks for the additional test, sorry the comments were still pending. This should be good now.

File: litex/soc/interconnect/axi.py [hunk 1]
```diff
@@ -182,10 +182,11 @@ def r_lite_description(data_width):
     ]
 
 class AXILiteInterface:
-    def __init__(self, data_width=32, address_width=32, clock_domain="sys", name=None):
+    def __init__(self, data_width=32, address_width=32, clock_domain="sys", name=None, bursting=False):
         self.data_width    = data_width
         self.address_width = address_width
         self.clock_domain  = clock_domain
+        self.bursting      = False  # Not supported in AXI Lite
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/axi.py]: Added.

File: litex/soc/interconnect/wishbone.py [hunk 2]
```diff
@@ -38,11 +39,17 @@
     ("err",              1, DIR_S_TO_M)
 ]
 
+CTI_BURST_NONE = 0b000
+CTI_BURST_CONSTANT = 0b001
+CTI_BURST_INCREMENTING = 0b010
+CTI_BURST_END = 0b111
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/wishbone.py]: Done.

File: litex/soc/interconnect/wishbone.py [hunk 3]
```diff
@@ -38,11 +39,17 @@
...
+
 
 class Interface(Record):
-    def __init__(self, data_width=32, adr_width=30):
-        self.data_width = data_width
-        self.adr_width  = adr_width
+    def __init__(self, data_width=32, adr_width=30, bursting=False):
+        self.data_width   = data_width
+        self.adr_width    = adr_width
+        self.bursting     = bursting
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/wishbone.py]: Fixed.

File: test/test_wishbone.py [hunk 4]
```diff
@@ -56,3 +56,44 @@ def __init__(self):
 
         dut = DUT()
         run_simulation(dut, generator(dut))
+
+    def test_sram_burst(self):
+        def generator(dut):
+            yield from dut.wb.write(0x0000, 0x01234567, cti=0b010)
+            yield from dut.wb.write(0x0001, 0x89abcdef, cti=0b010)
+            yield from dut.wb.write(0x0002, 0xdeadbeef, cti=0b010)
+            yield from dut.wb.write(0x0003, 0xc0ffee00, cti=0b111)
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply test/test_wishbone.py]: Done.

File: litex/soc/interconnect/wishbone.py [hunk 5]
```diff
@@ -346,6 +361,79 @@ def __init__(self, mem_or_size, read_only=None, init=None, bus=None):
...
+                                Cat(self.bus.adr & ~adr_wrap_mask[self.bus.bte],
+                                    self.bus.adr[adr_wrap_max:]
+                                    )
+                            ),
+                        ).Else(
+                            adr_counter.eq(
+                                Cat(self.bus.adr & ~adr_wrap_mask[self.bus.bte],
+                                    self.bus.adr[adr_wrap_max:]
+                                    ) + 1
+                            ),
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/wishbone.py]: Done, now `self.bus.we` is used for incrementing initial counter value before latching.

File: litex/soc/interconnect/wishbone.py [hunk 6]
```diff
@@ -357,17 +445,36 @@ def __init__(self, mem_or_size, read_only=None, init=None, bus=None):
...
+            self.bus.ack.eq(0)
         ]
+        if self.bus.bursting:
+            self.sync += [
+                If(self.bus.cyc & self.bus.stb & (~self.bus.ack | adr_burst), self.bus.ack.eq(1))
+            ]
+        else:
+            self.sync += [
+                If(self.bus.cyc & self.bus.stb & ~self.bus.ack, self.bus.ack.eq(1))
+            ]
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/wishbone.py:465]: Fixed.

File: litex/soc/interconnect/wishbone.py [hunk 7]
```diff
@@ -357,17 +445,36 @@ def __init__(self, mem_or_size, read_only=None, init=None, bus=None):
...
+                If(adr_burst & adr_latched,
+                    port.adr.eq(adr_next[:len(port.adr)]),
+                ).Else(
+                    port.adr.eq(self.bus.adr[:len(port.adr)]),
+                ),
+            ]
+        else:
+            self.comb += [
+                port.adr.eq(self.bus.adr[:len(port.adr)]),
+            ]
```
[the remark this answers is not in the document]
Comment (koluckirafal) [reply litex/soc/interconnect/wishbone.py]: Done.

**Review by enjoy-digital — APPROVED** [review]
Thanks @koluckirafal for the updates/changes. This now looks good and can be merged. I'll maybe do another review or some changes on the burst logic, now that this is integrated correctly and with tests, it will be be easy to do this.

**Comment by enjoy-digital** [comment]
Sorry, while reading the code it's not clear to me which exact type of bursts should be supported can you list them?. It seems that `adr_burst_wrap` is not used, so how is done the differentiation between a Incrementing Linear Burst and Wrapped Burst?

**Comment by koluckirafal** [comment]
Differentation between linear and wrapped incrementing address bursts is done by `BTE` signal - it's used to select the wrap mask, which then is being used for separating the first address into initial counter value and offset.
If `BTE` is equal `0`, then the first addess is loaded "as is" to the counter and offset register is unused; otherwise up to 4 least significant address bits (`BTE == 0b11`) are loaded to the offset register and are also set to `0` when setting initial counter value.
This way the same address counter logic can support both linear and {4, 8, 16}-beat wrapped increments.

`adr_burst_wrap` signal is probably a leftover from earlier experiments and can be removed.
